std/algorithm/iteration.d

```diff
@@ -3092,9 +3155,100 @@ if (isInputRange!RoR && isInputRange!(ElementType!RoR)
                 copy._items = _items.save;
                 copy._current = _current.save;
                 copy._currentSep = _currentSep.save;
+                static if (isBidirectional)
+                {
+                    copy._currentBack = _currentBack;
+                    copy._currentBackSep = _currentBackSep;
+                }
                 return copy;
             }
         }
+
+        static if (isBidirectional)
+        {
+            //no data in the last item of the initial range - use the separator
+            private enum useSepIfBackIsEmpty = q{
+                if (inputEndsWithEmpty)
+                {
+                    useBackSeparator;
+                    inputEndsWithEmpty = false;
+                }
+            };
+
+            private void setBackItem()
+            {
+                if (!_items.empty)
+                {
+                    _currentBack = _items.back.save;
+                }
+            }
+
+            private void useBackSeparator()
+            {
+                // Separator must always come after an item.
+                assert(_currentBackSep.empty && !_items.empty,
+                        "joiner: internal error");
```

Comment:
Please be more descriptive with the error. What should the user do when it sees "internal error"?

Reply:
I tried to be more descriptive.